File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    let results = [];
    let obj = {};
    let songInfo = {};
    
    for(let i = 0; i < genres.length; i++) {
        obj[genres[i]] = (obj[genres[i]] || 0) + plays[i];
        songInfo[i] = { play: plays[i], genre: genres[i]}
    }
    
    let sortedGenres = Object.keys(obj).sort((a, b) => obj[b] - obj[a]);
    
    sortedGenres.forEach((genre) => {
        const topSongs = Object.values(songInfo).
        filter((song) => song.genre === genre).
        sort((a, b) => b.play - a.play).
        slice(0, 2);
        
        topSongs.forEach((song) => {
            results.push(Object.values(songInfo).indexOf(song));
        })
    })
    
    
    
    return results;
}

/*

function solution(genres, plays) {

  sortedGenre.forEach((genre) => {
    // 장르별로 재생 횟수가 많은 노래 2개씩 추출
    const genreSong = Object.values(info)
      .filter((song) => song.genre === genre)
      .sort((a, b) => b.play - a.play)
      .slice(0, 2);

    // 고유 번호를 result에 담음
    genreSong.forEach((gs) => {
      const uniqueId = Object.values(info).indexOf(gs);
      result.push(uniqueId);
    });
  });

  return result;
}

*/

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    categories = {}
    play_count = []
    
    for i in range(len(genres)):
        genre = genres[i]
        play = plays[i]
        if not categories.get(genre):
            categories[genre] = {
                "total_play": 0,
                "play": []
            }
        categories[genre]["play"].append((play, i))
        categories[genre]["total_play"] += play
    
    for genre in categories.keys():
        play_count.append((categories[genre]["total_play"], genre))
        
    play_count.sort(key=lambda x:(-x[0]))
    print(play_count)
    
    for tplay, genre in play_count:
        categories[genre]["play"].sort(key=lambda x:-x[0])
        for play, idx in categories[genre]["play"][:2]:
            answer.append(idx)
    return answer
